### PR TITLE
Memorystat prep

### PR DESCRIPTION
**Pull request by JudahSchwartz**
## Contributing to Cortex XSOAR Content
Make sure to register your contribution by filling the [contribution registration form](https://forms.gle/XDfxU4E61ZwEESSMA)

**The Pull Request will be reviewed only after the contribution registration form is filled.**

## Status
- [ ] In Progress
- [x] Ready
- [ ] In Hold - (Reason for hold)

## Related Issues
fixes: XSUP-17806

## Description
mime_content is the actual file for the email, containing body/headers/attachments and all. We dont need this information if we arent saving the file. Removing this from the response dramatically reduces the data being sent.

The chunk_size parameter defines how many items should be loaded per fetch. The default is 100 but our default is 50. This means we will usually pull an extra 50 items we dont need. Setting the chunk_size to the fetch-limit should reduce the unneccessary data we pull

Notes for future improvements:
If memory is an issue while downloading attachments, o365 supports streaming file downloads, so we dont need to have the whole file in memory at a time
I noticed it sometimes will take up to 90 seconds simply to locate the folder on our environments. We must not be querying correctly/we should be caching the folder id or something

## Screenshots
Paste here any images that will help the reviewer

## Minimum version of Cortex XSOAR
- [ ] 6.0.0
- [ ] 6.1.0
- [ ] 6.2.0
- [ ] 6.5.0

## Does it break backward compatibility?
   - [ ] Yes
       - Further details:
   - [x] No

## Must have
- [ ] Tests
- [ ] Documentation

### CODE REVIEW

File: Packs/EWS/Integrations/EWSO365/EWSO365.py [hunk 1]
```diff
@@ -2030,7 +2032,8 @@ def parse_incident_from_item(item):
     """
     incident = {}
     labels = []
-
+    demisto.debug("starting parse to incident")
```
Comment (ShahafBenYakir) [inline Packs/EWS/Integrations/EWSO365/EWSO365.py:2035]:
```suggestion
    demisto.debug("starting to parse the email into an incident")
```

File: Packs/EWS/Integrations/EWSO365/EWSO365.py [hunk 2]
```diff
@@ -2195,7 +2204,11 @@ def parse_incident_from_item(item):
         labels.append({"type": "Email/ConversionID", "value": item.conversation_id.id})
 
     incident["labels"] = labels
+    demisto.debug("before generating rawJSON for incident")
+    log_memory()
     incident["rawJSON"] = json.dumps(parse_item_as_dict(item, None), ensure_ascii=False)
+    log_memory()
+    demisto.debug("after rawjson")
```
Comment (ShahafBenYakir) [inline Packs/EWS/Integrations/EWSO365/EWSO365.py:2211]:
```suggestion
    demisto.debug("FInshed generatiing rawjson")
```

File: Packs/EWS/Integrations/EWSO365/EWSO365.py [hunk 3]
```diff
@@ -2195,7 +2204,11 @@ def parse_incident_from_item(item):
         labels.append({"type": "Email/ConversionID", "value": item.conversation_id.id})
 
     incident["labels"] = labels
+    demisto.debug("before generating rawJSON for incident")
```
Comment (ShahafBenYakir) [inline Packs/EWS/Integrations/EWSO365/EWSO365.py:2207]:
```suggestion
    demisto.debug("Starting to generate rawJSON for incident")
```

File: Packs/EWS/Integrations/EWSO365/EWSO365.py [hunk 4]
```diff
@@ -2288,6 +2302,8 @@ def fetch_last_emails(
     :return: list of exchangelib.Items
     """
     qs = client.get_folder_by_path(folder_name, is_public=client.is_public_folder)
+    demisto.debug("finished folder get")
```
Comment (ShahafBenYakir) [inline Packs/EWS/Integrations/EWSO365/EWSO365.py:2305]:
```suggestion
    demisto.debug("Finished getting the folder by path")
```

File: Packs/EWS/Integrations/EWSO365/EWSO365.py [hunk 5]
```diff
@@ -2296,21 +2312,25 @@ def fetch_last_emails(
...
+    qs = qs.filter().only(*[x.name for x in Message.FIELDS if x.name != 'mime_content'])
     qs = qs.filter().order_by("datetime_received")
-
     result = []
     exclude_ids = exclude_ids if exclude_ids else set()
     demisto.debug(f'{APP_NAME} - Exclude ID list: {exclude_ids}')
+    qs.chunk_size = min(client.max_fetch, 100)
+    qs.page_size = min(client.max_fetch, 100)
+    demisto.debug("before iter")
+    demisto.debug(f'qs size:{sys.getsizeof(qs)}')
```
Comment (ShahafBenYakir) [inline Packs/EWS/Integrations/EWSO365/EWSO365.py:2323]: LEt's give a clearer log here

File: Packs/EWS/Integrations/EWSO365/EWSO365.py [hunk 6]
```diff
@@ -2296,21 +2312,25 @@ def fetch_last_emails(
...
     for item in qs:
+        demisto.debug("fetching an item")
```
Comment (ShahafBenYakir) [inline Packs/EWS/Integrations/EWSO365/EWSO365.py:2325]: let's give better log message here as well

File: Packs/EWS/Integrations/EWSO365/EWSO365.py [hunk 7]
```diff
@@ -2401,6 +2421,7 @@ def sub_main():
         elif command == "fetch-incidents":
             last_run = demisto.getLastRun()
             incidents = fetch_emails_as_incidents(client, last_run)
+            demisto.debug(f"returning incidents with size {sys.getsizeof(incidents)}")
```
Comment (ShahafBenYakir) [inline Packs/EWS/Integrations/EWSO365/EWSO365.py:2424]:
```suggestion
            demisto.debug(f"Saving incidents with size {sys.getsizeof(incidents)}")
```

File: Packs/EWS/Integrations/EWSO365/EWSO365.py [hunk 8]
```diff
@@ -2296,21 +2312,25 @@ def fetch_last_emails(
...
         qs = qs.filter(last_modified_time__gte=first_fetch_ews_datetime)
-    qs = qs.filter().only(*[x.name for x in Message.FIELDS])
+    qs = qs.filter().only(*[x.name for x in Message.FIELDS if x.name != 'mime_content'])
     qs = qs.filter().order_by("datetime_received")
-
     result = []
     exclude_ids = exclude_ids if exclude_ids else set()
     demisto.debug(f'{APP_NAME} - Exclude ID list: {exclude_ids}')
+    qs.chunk_size = min(client.max_fetch, 100)
+    qs.page_size = min(client.max_fetch, 100)
```
Comment (ShahafBenYakir) [inline Packs/EWS/Integrations/EWSO365/EWSO365.py:2321]: Does the user care about these defaults? (100)

**Review by ShahafBenYakir — COMMENTED** [review]
cool
Let's make sure when that when adding log messages we make them as clear as possible on what operation is being done such that future developers will have an easier time understanding what is happening.

File: Packs/EWS/Integrations/EWSO365/EWSO365.py (hunk 8 above)
Comment (JudahSchwartz) [reply Packs/EWS/Integrations/EWSO365/EWSO365.py:2321]: 100 is the given default at the moment from microsoft. I guess we can add parameters with warning signs for the customers. Should this be made into params?

Comment (ShahafBenYakir) [reply Packs/EWS/Integrations/EWSO365/EWSO365.py:2321]: How can it affect customers? Maybe just a disclaimer in docs

Comment (JudahSchwartz) [reply Packs/EWS/Integrations/EWSO365/EWSO365.py:2321]: this shouldnt affect customers, only performance. Page size and chunk size are instructions to the library for how to split up the http requests. Its an implementation detail
Maybe Im not understanding the question?

Comment (ShahafBenYakir) [reply Packs/EWS/Integrations/EWSO365/EWSO365.py:2321]: 👍

**Review by ShahafBenYakir — APPROVED** [review]

File: Packs/EWS/Integrations/EWSO365/EWSO365.py [hunk 9]
```diff
@@ -767,6 +767,8 @@ def parse_folder_as_json(folder):
         if type(value) in [str, str, int, float, bool, Body, HTMLBody, None]:
             raw_dict[field] = value
     raw_dict["id"] = item.id
+    demisto.debug("checking for attachments")
```
Comment (tkatzir) [inline Packs/EWS/Integrations/EWSO365/EWSO365.py:770]:
```suggestion
    demisto.debug(f"checking for attachments in email with id {item.id}")
```

File: Packs/EWS/Integrations/EWSO365/EWSO365.py [hunk 10]
```diff
@@ -2031,7 +2033,8 @@ def parse_incident_from_item(item):
     """
     incident = {}
     labels = []
-
+    demisto.debug("starting to parse the email into an incident")
```
Comment (tkatzir) [inline Packs/EWS/Integrations/EWSO365/EWSO365.py:2036]:
```suggestion
    demisto.debug(f"starting to parse the email with id {item.id} into an incident")
```

File: Packs/EWS/Integrations/EWSO365/EWSO365.py [hunk 11]
```diff
@@ -2069,7 +2072,11 @@ def parse_incident_from_item(item):
     # handle attachments
     if item.attachments:
         incident["attachment"] = []
+        demisto.debug(f"parsing {len(item.attachments)} attachments for item with id {item.id}")
+        attachment_counter = 0
         for attachment in item.attachments:
+            attachment_counter += 1
+            demisto.debug(f'retrieving attachment number {attachment_counter}')
```
Comment (tkatzir) [inline Packs/EWS/Integrations/EWSO365/EWSO365.py:2079]:
```suggestion
            demisto.debug(f'retrieving attachment number {attachment_counter} of email with id {item.id}')
```

File: Packs/EWS/Integrations/EWSO365/EWSO365.py [hunk 12]
```diff
@@ -2082,6 +2089,7 @@ def parse_incident_from_item(item):
 
                         # save the attachment
                         file_name = get_attachment_name(attachment.name)
+                        demisto.debug(f"saving content of size {sys.getsizeof(attachment.content)}")
```
Comment (tkatzir) [inline Packs/EWS/Integrations/EWSO365/EWSO365.py:2092]:
```suggestion
                        demisto.debug(f"saving content number {attachment_counter}, of size {sys.getsizeof(attachment.content)}, of email with id {item.id}")
```

File: Packs/EWS/Integrations/EWSO365/EWSO365.py [hunk 13]
```diff
@@ -2169,6 +2177,7 @@ def parse_incident_from_item(item):
             labels.append(
                 {"type": label_attachment_id_type, "value": attachment.attachment_id.id}
             )
+        demisto.debug(f'finished parsing attachment {attachment_counter}')
```
Comment (tkatzir) [inline Packs/EWS/Integrations/EWSO365/EWSO365.py:2180]:
```suggestion
        demisto.debug(f'finished parsing attachment {attachment_counter} of email with id {item.id}')
```

File: Packs/EWS/Integrations/EWSO365/EWSO365.py [hunk 14]
```diff
@@ -2196,7 +2205,11 @@ def parse_incident_from_item(item):
         labels.append({"type": "Email/ConversionID", "value": item.conversation_id.id})
 
     incident["labels"] = labels
+    demisto.debug("Starting to generate rawJSON for incident")
```
Comment (tkatzir) [inline Packs/EWS/Integrations/EWSO365/EWSO365.py:2208]:
```suggestion
    demisto.debug(f"Starting to generate rawJSON for incident, from email with id {item.id}")
```

File: Packs/EWS/Integrations/EWSO365/EWSO365.py [hunk 15]
```diff
@@ -2196,7 +2205,11 @@ def parse_incident_from_item(item):
...
+    log_memory()
     incident["rawJSON"] = json.dumps(parse_item_as_dict(item, None), ensure_ascii=False)
+    log_memory()
+    demisto.debug("FInshed generatiing rawjson")
```
Comment (tkatzir) [inline Packs/EWS/Integrations/EWSO365/EWSO365.py:2212]:
```suggestion
    demisto.debug(f"FInshed generatiing rawjson from email with id {item.id}")
```

File: Packs/EWS/Integrations/EWSO365/EWSO365.py [hunk 16]
```diff
@@ -2289,6 +2303,8 @@ def fetch_last_emails(
     :return: list of exchangelib.Items
     """
     qs = client.get_folder_by_path(folder_name, is_public=client.is_public_folder)
+    demisto.debug("Finished getting the folder by path")
```
Comment (tkatzir) [inline Packs/EWS/Integrations/EWSO365/EWSO365.py:2306]:
```suggestion
    demisto.debug(f"Finished getting the folder named {folder_name} by path")
```

File: Packs/EWS/Integrations/EWSO365/EWSO365.py [hunk 17]
```diff
@@ -2297,21 +2313,25 @@ def fetch_last_emails(
         assert first_fetch_datetime is not None
         first_fetch_ews_datetime = EWSDateTime.from_datetime(first_fetch_datetime.replace(tzinfo=tz))
         qs = qs.filter(last_modified_time__gte=first_fetch_ews_datetime)
-    qs = qs.filter().only(*[x.name for x in Message.FIELDS])
+    qs = qs.filter().only(*[x.name for x in Message.FIELDS if x.name != 'mime_content'])
```
Comment (tkatzir) [inline Packs/EWS/Integrations/EWSO365/EWSO365.py:2316]:
```suggestion
    qs = qs.filter().only(*[x.name for x in Message.FIELDS if x.name.lower() != 'mime_content'])
```

File: Packs/EWS/Integrations/EWSO365/EWSO365.py [hunk 18]
```diff
@@ -2297,21 +2313,25 @@ def fetch_last_emails(
...
     qs = qs.filter().order_by("datetime_received")
-
     result = []
     exclude_ids = exclude_ids if exclude_ids else set()
     demisto.debug(f'{APP_NAME} - Exclude ID list: {exclude_ids}')
+    qs.chunk_size = min(client.max_fetch, 100)
+    qs.page_size = min(client.max_fetch, 100)
+    demisto.debug("before iter")
```
Comment (tkatzir) [inline Packs/EWS/Integrations/EWSO365/EWSO365.py:2323]:
```suggestion
    demisto.debug("Before iterating on queryset")
```

**Review by tkatzir — COMMENTED** [review]
Great!
Some comments, mainly to add context to the log messages.

**Review by tkatzir — APPROVED** [review]
Great!

**Pull request merged**